tests/e2e/bq-batch-retrieval.py

```diff
@@ -392,3 +440,163 @@ def infra_teardown(pytestconfig, core_url, serving_url):
        print('Cleaning up not required')
+
+
+@pytest.fixture(scope="module")
+def update_featureset_dataframe():
+    n_rows = 10
+    time_offset = datetime.utcnow().replace(tzinfo=pytz.utc)
+    return pd.DataFrame(
+        {
+            "datetime": [time_offset] * n_rows,
+            "entity_id": [i for i in range(n_rows)],
+            "update_feature1": ["a" for i in range(n_rows)],
+            "update_feature2": [i + 2 for i in range(n_rows)],
+            "update_feature3": [i for i in range(n_rows)],
+            "update_feature4": ["b" for i in range(n_rows)],
+        }
+    )
+
+
+@pytest.mark.direct_runner
+@pytest.mark.run(order=20)
+def test_update_featureset_apply_featureset_and_ingest_first_subset(
+    client, update_featureset_dataframe
+):
+    subset_columns = ["datetime", "entity_id", "update_feature1", "update_feature2"]
+    subset_df = update_featureset_dataframe.iloc[:5][subset_columns]
+    update_fs = FeatureSet(
+        "update_fs",
+        entities=[Entity(name="entity_id", dtype=ValueType.INT64)],
+        max_age=Duration(seconds=432000),
+    )
+    update_fs.infer_fields_from_df(subset_df)
+    client.apply(update_fs)
+
+    client.ingest(feature_set=update_fs, source=subset_df)
+
+    time.sleep(15)
+    feature_retrieval_job = client.get_batch_features(
+        entity_rows=update_featureset_dataframe[["datetime", "entity_id"]].iloc[:5],
+        feature_refs=[
+            f"{PROJECT_NAME}/update_feature1",
+            f"{PROJECT_NAME}/update_feature2",
+        ],
+    )
+
+    output = feature_retrieval_job.to_dataframe().sort_values(by=["entity_id"])
+    print(output.head())
+
+    assert output["update_feature1"].to_list() == subset_df["update_feature1"].to_list()
+    assert output["update_feature2"].to_list() == subset_df["update_feature2"].to_list()
+
+
+@pytest.mark.direct_runner
+@pytest.mark.timeout(600)
+@pytest.mark.run(order=21)
+def test_update_featureset_update_featureset_and_ingest_second_subset(
+    client, update_featureset_dataframe
+):
+    subset_columns = [
+        "datetime",
+        "entity_id",
+        "update_feature1",
+        "update_feature3",
+        "update_feature4",
+    ]
+    subset_df = update_featureset_dataframe.iloc[5:][subset_columns]
+    update_fs = FeatureSet(
+        "update_fs",
+        entities=[Entity(name="entity_id", dtype=ValueType.INT64)],
+        max_age=Duration(seconds=432000),
+    )
+    update_fs.infer_fields_from_df(subset_df)
+    client.apply(update_fs)
+
+    # We keep retrying this ingestion until all values make it into the buffer.
+    # This is a necessary step because bigquery streaming caches table schemas
+    # and as a result, rows may be lost.
+    while True:
+        ingestion_id = client.ingest(feature_set=update_fs, source=subset_df)
+        time.sleep(15)  # wait for rows to get written to bq
+        rows_ingested = get_rows_ingested(client, update_fs, ingestion_id)
+        if rows_ingested == len(subset_df):
```

Comment:
If there is a transient failure, would we ever ingest the same rows twice? Should we dedupe in the select statement below perhaps?

Reply:
I think its enough of an edge case to ignore it in this query, especially if we want to keep it cheap and simple